### PR TITLE
Render personal_sign messages as utf-8 text

### LINKED ISSUE #1173: personal_sign - render message to sign as utf8

### PR DESCRIPTION
**Pull request by danfinlay**
Calls to `personal_sign` are now:

- When hex encoded, preserved as hex encoded, but displayed as utf-8 text.
- When not hex encoded, decoded as utf-8 text as hex for signing.
- The messages proposed for signing are displayed as UTF-8 text.
- When the message cannot be rendered as UTF-8 text, it is displayed as hexadecimal.

screen shot 2017-03-06 at 2 58 55 pm

Fixes #1173

### CODE REVIEW

**Comment by kumavis** [comment]
> When not hex encoded

I think we should hard fail here, as hex encoding is a valid utf8 encoding

**Comment by flyswatter** [comment]
Sample exception: `0x1` is valid hex but is not utf8 decodable.

**Comment by flyswatter** [comment]
What would a hard fail look like here? Throw an error to the sign request, and not show it to the user?

**Comment by kumavis** [comment]
throw an error to the sign request

**Comment by kumavis** [comment]
One example of unexpected behavior would be signing a number (e.g. a unix timestamp) -- in this implementation it would be interpreted as hex instead of utf8

File: ui/app/components/pending-personal-msg-details.js [hunk 1]
```diff
@@ -21,6 +22,7 @@ PendingMsgDetails.prototype.render = function () {
   var account = state.accounts[address] || { address: address }
 
   var { data } = msgParams
+  console.dir({ msgParams })
```
Comment (kumavis) [inline ui/app/components/pending-personal-msg-details.js]: plz remove

**Comment by flyswatter** [comment]
The old `eth_sign` implementation hard-requires a hex-prefix `0x`, and otherwise interpretes as utf8. That also seemed weird. What if a utf8 string started with `0x`?

The safest move would maybe be to only allow hex input, try to decode it, but fall back to the original input.

**Review by kumavis — CHANGES_REQUESTED** [review]

**Comment by kumavis** [comment]
> The safest move would maybe be to only allow hex input, try to decode it, but fall back to the original input.

this isnt safe if the developer is expecting the fallback case but (accidently) ends up with valid hex

**Comment by flyswatter** [comment]
This is what MIME types are for, and what I was originally postponing this feature for.  Maybe we should just wait for that, I don't see a correct way forward here.

**Comment by kumavis** [comment]
my philosophy here is:
for developer input, have one correct way of doing things
for user input, be flexible

**Comment by kumavis** [comment]
I look forward to the MIME type spec - but that feature actually happens "after" this intial decoding

we are dealing with the first layer of this stack:
- decode raw RPC params
- parse for MIME type
- display message signing UI

**Review by kumavis — APPROVED** [review]

**Comment by chafey** [comment]
In many cases, a timestamp will be included in the message to be signed to prevent replay attacks.  The timestamp is not necessarily something we want to display to the user though - yet this scheme would do so.  I would love to be able to pass a "display string" that is presented to the user in the confirmation window to help guide them through the process.  For example "Please press "Confirm" if you would like to do X".  We can make the actual message available to the user for inspection if they so desire, but it may not be meaningful to them

**Comment by flyswatter** [comment]
I think mime content headers would adequately allow specifying portions of the signed text to render vs to hide: https://github.com/MetaMask/metamask-plugin/issues/925

**Pull request merged**